### DISCUSSION TITLE
Test Automation -WebView2 - [ Selenium + MSEdge Driver + Java ] Exception

### DISCUSSION THREAD

**Discussion by zsmjluo** [MicrosoftEdge/WebView2Feedback#3892] [Q&A]
Use java not c#

[WebView2 APP start use "--remote-debugging-port=9222"]

`System.setProperty("webdriver.edge.driver", "D:\\\\msedgedriver.exe");
ChromeOptions chromeOptions = new ChromeOptions();
chromeOptions.setExperimentalOption("debuggerAddress", "127.0.0.1:9222");
		    
EdgeOptions options = new EdgeOptions();
options.setCapability("UseWebView", true);
options.setCapability("UseChromium", true);

EdgeDriverService service = EdgeDriverService.createDefaultService();
EdgeDriver driver = new EdgeDriver(service, options.merge(chromeOptions));`

But I get Exception:
_session not created: No matching capabilities found
Build info: version: 'unknown', revision: 'unknown', time: 'unknown'_

**Comment by champnic** [comment]
Hey @zsmjluo - At what point in the code do you get the error?

**Comment by zsmjluo** [comment]
> Hey @zsmjluo - At what point in the code do you get the error?

@champnic 
This code reports an exception：

“EdgeDriver driver = new EdgeDriver(service, options.merge(chromeOptions));“


org.openqa.selenium.SessionNotCreatedException: session not created: No matching capabilities found
Build info: version: 'unknown', revision: 'unknown', time: 'unknown'
System info: host: '', ip: '', os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '1.8.0_231'
Driver info: driver.version: EdgeDriver
remote stacktrace: Backtrace:
	Ordinal0 [0x00F1D8C3+3594435]
	Ordinal0 [0x00C33884+538756]
	Ordinal0 [0x00C6FC40+785472]
	Ordinal0 [0x00C6FFC6+786374]
	Ordinal0 [0x00C6FDDE+785886]
	Ordinal0 [0x00C6C64B+771659]
	Ordinal0 [0x00C51FED+663533]
	Ordinal0 [0x00C52C8A+666762]
	Ordinal0 [0x00C52C27+666663]
	Ordinal0 [0x00E23E5C+2571868]
	GetHandleVerifier [0x010391CA+1051786]
	GetHandleVerifier [0x01046FA4+1108580]
	GetHandleVerifier [0x01039C91+1054545]
	Ordinal0 [0x00E1B7AE+2537390]
	Ordinal0 [0x00E25C2B+2579499]
	Ordinal0 [0x00E25DC1+2579905]
	Ordinal0 [0x00E374F5+2651381]
	BaseThreadInitThunk [0x76B56359+25]
	RtlGetAppContainerNamedObjectPath [0x77C27C24+228]
	RtlGetAppContainerNamedObjectPath [0x77C27BF4+180]

Could you give me a correct Java code example to test webView2 APP？Thanks

**Comment by champnic** [comment]
@zsmjluo What version of Selenium are you using?

**Comment by zsmjluo** [comment]
> @zsmjluo What version of Selenium are you using?

@champnic selenium 3

**Comment by bwalderman** [comment]
Hi @zsmjluo. Mixing ChromeOptions and EdgeOptions using is not supported and will not result in the correct capabilities being passed to msedgedriver.exe. Also, the Edge classes in Selenium 3 are out-of-date and do not support Edge Chromium or WebView2.

To automate WebView2 with Java, please consider upgrading to Selenium 4 which supports Edge Chromium and WebView2 natively.

If you want to continue using Selenium 3, you will need to add the Selenium Tools for Microsoft Edge to your project. You can find instructions for using the Tools with Java at https://github.com/microsoft/edge-selenium-tools#java. The following sample should get you started successfully automating WebView2 with Selenium 3 + Selenium Tools for Microsoft Edge. First import the Edge classes from our library:

```java
import com.microsoft.edge.seleniumtools.EdgeDriver;
import com.microsoft.edge.seleniumtools.EdgeDriverService;
import com.microsoft.edge.seleniumtools.EdgeOptions;
```

Then, use EdgeOptions and EdgeDriver normally. These Edge classes imported from seleniumtools support all of the new Chromium options:

```java
System.setProperty("webdriver.edge.driver", "D:\\msedgedriver.exe");

EdgeOptions options = new EdgeOptions();
options.setExperimentalOption("debuggerAddress", "127.0.0.1:9222");

EdgeDriverService service = EdgeDriverService.createDefaultService();
EdgeDriver driver = new EdgeDriver(service, options);
```

Note that useChromium is not necessary because seleniumtools uses Chromium by default. UseWebView is also not necessary because this is only needed if you want msedgedriver to launch your WebView2 app for you. Since you are attaching using debuggerAddress in your example, I assume you are already launching your app some other way.
